File: LeetCode/0045.medium.跳跃游戏II.py
# -*- encoding: utf-8 -*-
'''
@File    :   0045.medium.跳跃游戏II.py
@Time    :   2022/06/26 16:14:05
@Author  :   Zhifeng Li
@Contact :   li_zaaachary@163.com
@Desc    :   
'''
from typing import List


class Solution_greed:
    '''
    贪心算法，和dp同理，但是不需要矩阵
    '''
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        maxPos, end, step = 0, 0, 0
        for i in range(n - 1):
            if maxPos >= i:
                maxPos = max(maxPos, i + nums[i])
                if i == end:
                    end = maxPos
                    step += 1
        return step


class Solution_dp_v2:
    '''
    dp v2 删除部分操作 执行用时： 56 ms , 在所有 Python3 提交中击败了 51.33% 的用户 内存消耗： 15.9 MB , 在所有 Python3 提交中击败了 26.67% 的用户
    '''
    def jump(self, nums: List[int]) -> int:
        dp = [i for i in range(len(nums))]
        max_idx = 0

        for idx in range(len(nums)):
            temp = idx + nums[idx]
            if max_idx < temp:
                for jdx in range(max_idx, temp+1):
                    if jdx < len(nums):
                        dp[jdx] = min(dp[idx]+1, dp[jdx])
                    else:
                        break
                max_idx = temp

        return dp[-1]


class Solution_dp_v1:
    '''
    dp v1 超时
    '''
    def jump(self, nums: List[int]) -> int:
        dp = [i for i in range(len(nums))]
        max_idx = 0

        for idx in range(len(nums)):
            max_idx = max(max_idx, idx + nums[idx])
            
            for jdx in range(idx+1, max_idx+1):
                if jdx < len(nums):
                    dp[jdx] = min(dp[idx]+1, dp[jdx])
                else:
                    break

        return dp[-1]

if __name__ == "__main__":
    S = Solution()
    print(S.jump([2,3,1,1,4]))

File: LeetCode/0045.medium.跳跃游戏II.cpp
/*
@File    :   0045.medium.跳跃游戏II.cpp
@Time    :   2022/06/26 15:52:04
@Author  :   Zhifeng Li
@Contact :   li_zaaachary@163.com
@Desc    :   
给你一个非负整数数组 nums ，你最初位于数组的第一个位置。

数组中的每个元素代表你在该位置可以跳跃的最大长度。

你的目标是使用最少的跳跃次数到达数组的最后一个位置。

假设你总是可以到达数组的最后一个位置。

*/
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        vector<int> dp(nums.size());
        for (int i = 0; i < dp.size(); i++)
            dp[i] = i;

        int max_idx = 0;

        for (int idx = 0; idx < nums.size(); idx ++){
            int temp = idx + nums[idx];
            if (max_idx < temp){
                for (int jdx = max_idx; jdx < temp +1; jdx++ ){
                    if (jdx < nums.size())
                        dp[jdx] = dp[jdx] > dp[idx] + 1 ? dp[idx] + 1 : dp[jdx];
                    else
                        break;
                }
                max_idx = temp;
            } 
        }
        
        return *(dp.end() - 1);
    }
};

int main(int argc, char const *argv[])
{
    Solution S;
    vector<int> nums{2,3,1,1,4};
    cout << S.jump(nums);
    return 0;
}
